tests/purs/failing/FoldableInstance4.out

```diff
@@ -0,0 +1,20 @@
+Error found:
+in module [33mFoldableInstance4[0m
+at tests/purs/failing/FoldableInstance4.purs:8:1 - 8:27 (line 8, column 1 - line 8, column 27)
+
+  The following data constructors and their subterms must use the type variable 'a' only as the last argument of a data type:
+
+    Constructor, T has invalid 'a' usage at path(s):
+
+      _0\out <functions are not supported for this type class>
```

Comment:
I don't think we should have a `Function`-specific error message for this. `Function t` is not `Foldable`, so this should be the same message that any other uses-`a`-but-not-`Foldable` type appearing in the `T` constructor would produce. (If that error message is inadequate, let's fix it for all not-`Foldable` types.)

Reply:
I'm still treating Functions specially, but that is something that can easily be changed with the `InvalidConstructorArgReason` type. Do you know of other examples where we have an error like that, so I can reuse its phrasing?

Reply:
We might be talking past each other. Either:
* Don't handle functions specially. The compiler will eventually raise a `NoInstanceFound` when it tries to `foldMap` or `traverse` over a function, with a message along the lines of ‘No type class instance was found for Foldable (Function t)’, which IMO is a fine error message for that case and matches what happens if any other constructed type that isn't `Foldable`/`Traversable` is used incorrectly.
* Don't handle functions specially, and check up front that any constructed types that need to be folded/traversed over implement the right class. If they don't, the message used for that sort of `InvalidConstructorArgReason` should be similar to the message for `NoInstanceFound`, with whatever additional information would make this option a better choice than the first.

Reply:
Ah... I see now. I forgot that `Function` is subsumed by the recursive arg handling since we're no longer supporting contravariant positions.